File: solutions/nim/bob/1/bob.nim
import std/strutils

proc ends_with_questionmark(s: string): bool =
  let stripped = strip(s)
  return (stripped.len > 0) and (stripped[^1] == '?')

proc is_yelled(s: string): bool =
  return s.contains(UppercaseLetters) and (s.toUpperAscii == s)

proc hey*(s: string): string =
  if not is_yelled(s) and ends_with_questionmark(s):
    return "Sure."
  if is_yelled(s) and  not ends_with_questionmark(s):
    return "Whoa, chill out!"
  if is_yelled(s) and ends_with_questionmark(s):
    return "Calm down, I know what I'm doing!"
  if isEmptyOrWhitespace(s):
    return "Fine. Be that way!"
  else:
    return "Whatever."

File: solutions/ocaml/bob/1/bob.ml
let response_for text =
  let s = String.trim text in
  let is_empty_trimmed = String.length  s = 0 in
  let is_uppercase = (String.uppercase_ascii s) = s && (String.uppercase_ascii s) <> (String.lowercase_ascii s) in
  let is_question = (String.ends_with ~suffix:"?" s) in
  match (is_empty_trimmed, is_uppercase, is_question) with
  | (true, _, _) -> "Fine. Be that way!"
  | (false, true, false) -> "Whoa, chill out!"
  | (false, true, true) -> "Calm down, I know what I'm doing!"
  | (false, false, true) -> "Sure."
  | _ -> "Whatever."
